kroxylicious/src/main/java/io/kroxylicious/proxy/frame/DecodedRequestFrame.java

```diff
@@ -19,22 +19,51 @@
 import org.apache.kafka.common.message.RequestHeaderData;
 import org.apache.kafka.common.protocol.ApiMessage;
 
+import io.kroxylicious.proxy.future.ProxyPromise;
+
 /**
  * A decoded request frame.
  */
 public class DecodedRequestFrame<B extends ApiMessage>
         extends DecodedFrame<RequestHeaderData, B>
         implements RequestFrame {
 
-    private final boolean decodeResponse;
+    private final ProxyPromise<? extends Object> promise;
+    boolean decodeResponse;
+    private final Object recipient;
 
-    public DecodedRequestFrame(short apiVersion,
-                               int correlationId,
-                               boolean decodeResponse,
-                               RequestHeaderData header,
-                               B body) {
+    private DecodedRequestFrame(short apiVersion,
+                                int correlationId,
+                                boolean decodeResponse,
+                                Object recipient,
+                                ProxyPromise<? extends Object> promise,
+                                RequestHeaderData header,
+                                B body) {
         super(apiVersion, correlationId, header, body);
+        if (recipient != null && !decodeResponse) {
+            throw new IllegalArgumentException();
+        }
         this.decodeResponse = decodeResponse;
+        this.recipient = recipient;
+        this.promise = promise;
+    }
+
+    public static <B extends ApiMessage> DecodedRequestFrame<B> internalRequest(short apiVersion,
+                                                                                int correlationId,
+                                                                                boolean decodeResponse,
+                                                                                Object recipient,
+                                                                                ProxyPromise<? extends Object> promise,
+                                                                                RequestHeaderData header,
+                                                                                B body) {
+        return new DecodedRequestFrame<B>(apiVersion, correlationId, decodeResponse, recipient, promise, header, body);
+    }
+
+    public static <B extends ApiMessage> DecodedRequestFrame<B> clientRequest(short apiVersion,
```

Comment:
Not sure these static factories bring much, but they do separate the difference cases.